### DISCUSSION TITLE
Production Website: Best Practice Question related to lifecycle startup/shutdown and async sql template

### DISCUSSION THREAD

**Discussion by VirajVaitha123** [fastapi/fastapi#11927] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
//router example.py
from fastapi import APIRouter,HTTPException, Request, Depends
from scryptify.logger import get_logger

logger = get_logger(__file__)

# Initialize the FastAPI application
router = APIRouter()
# Define the cleanup function

def get_sql_handler(request: Request):
    return request.app.state.sql_handler


@router.delete("/delete/{user_id}/{file_id}/{playlist_id}")
async def delete_from_playlist(user_id: str, file_id: str, playlist_id: str,sql_handler = Depends(get_sql_handler)):
    """
    Deletes a file entry from the database based on user_id and file_id.
    """
    try:
        # Call the delete_row function from SQLHandler
        sql_handler.delete_from_playlist(user_id, file_id,playlist_id)
        return {"message": "File successfully deleted"}
    except Exception as err:
        logger.error("An error occurred during file deletion: %s", str(err))
        raise HTTPException(status_code=500, detail="Internal Server Error") from err

//main.py
from dotenv import load_dotenv
import sys
import os
from fastapi import FastAPI
from contextlib import asynccontextmanager
from routers import delete_endpoints
from routers import eventgrid_router
from scryptify.logger import get_logger
from scryptify.cloud_manager import BlobHandler, SQLHandler

logger = get_logger(__name__)
load_dotenv() 

@asynccontextmanager
async def lifespan(application: FastAPI):
    """
    Initializes global resources at startup to improve request latency and resource efficiency.
    Loading data and initializing services once reduces redundant tasks for each request.
    Attributes:
        application.state.data_handler: Manages data operations.
        application.state.pdf_service: Handles PDF-related operations.
    """
    logger.info("🚀 Loading Global Resources...")
    application.state.blob_handler = BlobHandler()
    application.state.sql_handler = SQLHandler()
    yield
    logger.info("🌟 Cleaning up Global Resources...")
    if hasattr(application.state, 'blob_handler'):
        application.state.blob_handler = None
    if hasattr(application.state, 'sql_handler'):
        application.state.sql_handler = None

app = FastAPI(......#not relevant hidden)

# Import gallery service
app.include_router(delete_endpoints.router)
app.include_router(eventgrid_router.router)
```

### Description

Hi,

My service uses a event driven microservice architecture. One component of my service, is a FastAPI app, where some of the tasks involve interacting with SQL and BlobStorage.

I am in the stage of revisiting all of my microservices, and ensuring they are surely async, and optimized as much as possible. It's also a good chance for me to upskill.

**Question 1:**
What is the recommended practice in instantiating a SQLhandler class in FastAPI apps. On start up or inside the endpoint. My sql_handler = SQLHandler() requires no params, it simply connects to the SQL database. I have provided more context below, at a high level I thought it was best on start up and think I was proved wrong.

I used the lifecycle context manager to instantiate once on start up since it seems like a good idea not to reconnect sql everytime a request is made.

However, my azure log stream is showing the SQL connection fails after some time. I suppose the connection times out when idle, or after the shutdown process occurs.

 I could place some logic that checks the connection and if not reconnects it, but then I thought why not just initiate the handler when the API is called. The BlobHandler seemed fine for some reason and didn't really disconnect.

**Question 2: SQL for a website that could be serving multiple users should be async?**
There is a async approach to sql, and I was wondering if you all think it is necessary. I've had multiple users use the website and it seems like everything works fine. If not, are their recommended templates or examples (detailed) since it must be a common usecase of developers using FastAPI to interact with sql. ChatGPT sometime spits out nonsense, so I just want to be sure.

### Operating System

Windows

### FastAPI Version

doesnt matter

### Pydantic Version

doesnt matter

### Python Version

doesnt matter

**Comment by YuriiMotov** [comment, marked as answer]
We don't see the implementation of your classes. So, I can't say for sure that it will work correctly.

Assuming the implementation of your `SQLHandler` includes session pool management logic (or session management without pool), it should work.
But I would split session management logic from repository logic. It would give you more freedom of transaction management (perform several DB operations in one transaction).

Regarding the mentioned error, it's hard to say something not seeing the implementation of `SQLHandler`.

What is not OK is that you are using sync IO-methods inside the async endpoints. This will cause performance issues since your sync DB operations will block async loop every time you interact with DB.

Regarding Question 2: it's not necessary to use async. Async approach helps you use your hardware in a more efficient way (CPU will not just wait for the answer from DB, but will switch to another async task while waiting. So, you can handle more parallel request with the same CPU). But if you don't care about the efficiency of using your CPU, you can use sync approach and scale your system by adding more workers as you need it.
